src/lib.rs

```diff
@@ -105,39 +106,56 @@ pub struct Kaboodle {
     broadcast_port: u16,
     self_addr: Option<SocketAddr>,
     cancellation_tx: Option<Sender<()>>,
+    interface: Interface,
 }
 
 impl Kaboodle {
     /// Create a new Kaboodle mesh client, broadcasting on the given port number. All clients using
     /// a given port number will discover and coordinate with each other; give your mesh a distinct
     /// port number that is not already well-known for another purpose.
-    pub fn new(broadcast_port: u16) -> Kaboodle {
+    pub fn new(
+        broadcast_port: u16,
+        preferred_interface: Option<Interface>,
+    ) -> Result<Kaboodle, KaboodleError> {
         // Maps from a peer's address to the known state of that peer. See PeerState for a
         // description of the individual states.
         let known_peers: HashMap<Peer, PeerState> = HashMap::new();
 
-        Kaboodle {
+        let Some(interface) = preferred_interface.or_else(|| {
+            // If no interface was provided, use the first IPv6 interface we find, and if there are
+            // no IPv6 interfaces, use the first IPv4 interface.
+            let (ipv6interfaces, ipv4interfaces): (Vec<Interface>, Vec<Interface>) = non_loopback_interfaces()
+                .into_iter()
+                .partition(|iface| matches!(iface.addr, IfAddr::V6(_)));
+
+            ipv6interfaces.get(0).or_else(|| ipv4interfaces.get(0)).cloned()
```

Comment:
I feel like there oughta be a clearer way to express this, and I am open to feedback.

Reply:
This did work (modulo some braces and borrow-checker-satisfaction)! Much nicer. I ended up being inspired by your change and came up with a slightly more compact representation that I like:

```rust
let non_loopbacks = non_loopback_interfaces();
            let first_ipv6_interface = non_loopbacks.iter().find(|xs| matches!(xs.addr, IfAddr::V6(_)));
            first_ipv6_interface.or_else(|| non_loopbacks.first()).cloned()
```